### DISCUSSION TITLE
Ability to convert units between metric / imperial / US imperial

### DISCUSSION THREAD

**Discussion by enrauk** [mealie-recipes/mealie#3881] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/mealie-recipes/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

Allow users to toggle between metric / imperial / US imperial units for a recipe and automatically convert the units to the chosen system.  This could either be a profile setting or per recipe (per recipe would be more flexible)

### Please provide a concise description of the feature that would resolve your issue.

This would greatly assist users when importing recipes that use a different measurement system to their chosen preference.

### Please consider and list out some caveats or tradeoffs made in your design decision

Some measurements may not cover easily such as cups; most recipe managers do not attempt to convert these.

### Additional Information

- [ ] If this is accepted I'm willing to submit a PR to provide this feature
- [ ] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by ronnybugle** [comment]
This would be so awesome!

**Comment by TheCodeTrainer** [comment]
This would be an incredible feature!

**Comment by Spinkman** [comment]
if you're going to do this, please separate volume and temperature measurements. (I like my temps in F and measurements in metric)

**Comment by Yehuda-Bergstein** [comment]
This would be great!

**Comment by tecosaur** [comment]
It occurs to me that a neat way of doing this might be by introducing a concept of "canonical units". By this, I mean the ability to say that the "canonical" form of `1 lb` is `0.454 kg`.

These transformations could then be applied when importing/viewing a recipe.

**Reply by tecosaur** [reply]
I'd imagine this could also help with regional differences in spelling, e.g. `liter` vs `litre`.

**Reply by tit1** [reply]
Im working on a recipe system and the way i handle it base units are either grams or milliliter. All measurements compute down to those base measurement and than back up to the desired measurement format.

**Reply by Ronaldvr** [reply]
Actually this *is* already the *definition* all 'imperial units' are now defined by metric ones:

Since the [Weights and Measures Act 1985](https://en.wikipedia.org/wiki/Weights_and_Measures_Act_1985), British law defines base imperial units in terms of their metric equivalent.

https://en.wikipedia.org/wiki/Imperial_units

and
The majority of U.S. customary units were redefined in terms of the [meter](https://en.wikipedia.org/wiki/Metre) and [kilogram](https://en.wikipedia.org/wiki/Kilogram) with the [Mendenhall Order](https://en.wikipedia.org/wiki/Mendenhall_Order) of 1893 and, in practice, for many years before.[[3]](https://en.wikipedia.org/wiki/United_States_customary_units#cite_note-Mendenhall-3) These definitions were refined by the [international yard and pound](https://en.wikipedia.org/wiki/International_yard_and_pound) agreement of 1959.[[4]](https://en.wikipedia.org/wiki/United_States_customary_units#cite_note-FR59-5442-4)
https://en.wikipedia.org/wiki/United_States_customary_units

**Comment by EinToni** [comment]
This would be a great feature! But I guess for this to work, mealie would have to switch from using a plane string as a unit, to defining actual units. This would also open the door for many other features.

I already don't understand why you can input anything as a unit. There only exist two systems: metric and imperial, so why would one allow the use of anything when 99% of the people just fill in the units with the same metric and imperial names? So changing the units to predefined ones would be only reasonable, I think.

Maybe some mealie contributor will read this :) I would be happy to help implementing this if there is the will to move in that direction.

**Comment by enrauk** [comment]
Just to note that British, Australian, and American Imperial systems differ
on some measures.  A tablespoon in the UK is equivalent to 15 ml (same as
the EU) whereas in the Australia it is 20 ml and in the US 14.8 ml,
for example.

Regards, Enrico

**Comment by Christian-Me** [comment]
Take a look into this [wikipedia article](https://en.m.wikipedia.org/wiki/Cooking_weights_and_measures)
I stumbled upon this while writing a converter transferring my recipes over to mealie.

It mentions metric units in one sentence 

`In most of the world, recipes use the metric system of units—litres (L) and millilitres (mL), grams (g) and kilograms (kg), and degrees Celsius (°C).`

Saying that most of the world uses 3! Units: grams, liters, and degrees Celsius. That's it, everything else is confusing.🫤

Sure we use convenient things like tablespoons or teaspoons too. But containers, cans or bottles might be wrong if the food industry keeps on downsizing everything.

Perhaps scale everything internally down to metric. From there the appropriate unit could be scaled by users preferences and value (i.e. if more than 1000g than use kg)
Leave special cooking units like teaspoons or tablespoons untouched as they vary by ingredient.

User preferences could be selected differently for weight, volume and temperature to be able to use Degrees Fahrenheit - for whatever reason 😂
Or even single specific preference for individual food items.

Than nutrition values could be calculated automatically out of the ingredients. And perhaps price (it's getting more and more interesting) in future.

**Reply by fbd-ss** [reply]
most of the world being the ones embracing change, geez even the brits converted ; )

**Comment by Spinkman** [comment]
So, in my own cooking process, I prefer metric and tend to eyeball things. 
that being said, and this may be a separate feature request, it would be really nice to have versioning within the recipe. 
I'll make the recipe once, and often make changes on the fly based on my family's tastes.
my current work flow is the edit the mealie recipe manually with the changes I did and maybe leave some notes.

Now, how this can apply to unit conversion is to just leave it up to the person if they want to do any conversions at all.
when importing, Mealie can try to determine what unit of measure is being used so that conversions could be done in the future. currently ingredients are all bunched together and aren't numbers/units. so this would be the first (and most painful) step.

After there are separate DB entries for amounts and units separately, then you could have options to define if you want to use automatic unit conversion and also have an area where the user can define the conversion they'd like to use (eg: teaspoon = 14.8? or round to 15?)
doing it this way, you can define for yourself that you'd like a teaspoon to be 5ml a tablespoon to be 15ml and a cup to be 250ml. (this is what I use in my head because it's easier math.)

leaving it as defaults still gets you conversions if you choose and I would suggest rounding the numbers to the nearest integer. or base 5/10

anyway, more options for unit localization only serve to increase world wide usage of this already very useful program.

**Comment by enrauk** [comment]
Cup measures are also different UK and AU is 250 ml same as metric but US
is 240 ml, so I'd convert to 250 ml.  Teaspoons and tablespoons if you
wanted to convert them I would use metric measures 5 ml (UK, AU and metric
5ml, US 6 ml) and 15 ml respectively (UK and metric 15 ml, US 14.7 ml, AU
20 ml).  The differences for most of the world aren't great.

Fluid oz to ml I would use 29.57 (US and AU 29.57, UK 28.41).

One other point to note is that fl oz to pint is different: UK and AU it's
20 fl oz but for the US it's 16.

I can't think of any others that you might need to know.

Regards, Enrico

**Reply by Ronaldvr** [reply]
Which of course is actually a definitive reason to *not* use ambiguous measurement definitions as a starting point, but just metric which is the same everywhere.

**Comment by ronnybugle** [comment]
I use metric (Aus), but I use cups, teaspoons and tablespoons because it's easier to use a measuring cup set.  I'm not that bothered whether the recipe is North American and isn't exactly the same measurements, it seems close enough.  I would only want to convert ingredients that are like a 14oz can of whatever or 4 oz of mushrooms, because oz and fl oz mean nothing to me.

**Comment by CloCkWeRX** [comment]
https://github.com/hgrecco/pint is a possibility.
Not sure if other packages exist for customary units

**Comment by dokeraj** [comment]
I would also like to see this feature be implemented - very much needed.

**Comment by ph0b0s101** [comment]
would be great to see this feature in one of the upcoming releases.

**Comment by AlexandruNegura** [comment]
I recently installed Mealie and this is the last missing feature that I want. 
The AI parser is great, cutting the servings is GODLIKE.

But this conversion between metric types seems an easy win :D

**Comment by OlliWare** [comment]
Boosting this too, this is really needed in Mealie!

**Comment by ArLo69** [comment]
I would like to suggest e flexible setup unit of measure system.
My idea is based on what Dynamics 365 Business Central does https://www.youtube.com/watch?v=NzHTbPXXEr8
Only here the base unit of measure is registered in the item and other options are derived from there. Note that wehile watching this video.
We might do a simple version which I will explain in the metric system (and some Dutch units of measure)

We make a table with all units of measure. Unit is the name of the unit of measure itself. Base UoF is the Unit of Measure to calculate to. Quantity is the qunatity this unit has of the base Unit of Measure
```
Unit       Base UoM     Quantity
gram       -empty-       1 (should allways be 1 for base UoM)
kilogram   gram          1000
pond       gram          500 (a Dutch name for 500 grams)
ons        gram          100 (a Dutch name for 100 grams)
centiliter -empty-       1
liter      centiliter    100
deciliter  centiliter    10
```

One can set up this table for what he needs and use metric or imperial as base units.
Then you can have imperial to metric for example
```
oz      centiliter      2.957 (decide for yourself how to round)
gallon  centiliter      378.541
ounce   gram            28.35
pound   gram            453.592
```
On the ingredients you can set in which Unit of Measure you buy it in a field like Purchase UoM
For me that would be per Pond or kilogram for fruit and vegetables.
For canned food that might be Can500gr or Can400gr or Can250gr
In the Unit Of Measure table we can define that like:
```
Can250gr    gram    250
Can400gr    gram    400
Can500gr    gram    500
```

With a Unit of Measure table like this each user can set it up for his own needs. You can convert from imperial to metric, but are also free to calculate between any unit of measure as long as you define a relation between them.
This can be used to create shopping lists for example when there are more recipes with the same ingredient.

I hope my idea helps to set up something usefull. I am willing to think along and test if needed.

**Reply by crimsoncor** [reply]
In  a sense I think it is even simpler than that (kinda until I go and add even more complexity).  You technically don't even need the empty entries, just a list of conversions from one unit to another.  And it doesn't have to be hard-coded at all, though having Mealie ship with a list of "standard" conversions that the users can load if they want is a good idea. 

This is also very useful outside of metric/imperial when using imperial volume measurements. I might be using a recipe that calls for 4 tablespoons but when I scale it up 3x, I'd rather see 3/4 cup since I'm going to use a large measuring cup to measure it vs a small tablespoon. But that is also a slightly different problem and can get weird because if I end with a recipe calling for 11 tbsp, I don't really want to see 0.6875 cups. What I'd love to see is 1/2 cup + 3 tbsp (this gets back to the request raised in this ticket https://github.com/mealie-recipes/mealie/discussions/3127). 

But I do think there might be a solution that can handle all of these things using the same basic conversion table as the underlying data source. After that it is simply an issue of how much effort people want to put into setting up the relationships.

So the data source is just unit to unit values 
```
1    oz     28.3  grams
1    tbsp   3     tsp 
1000 grams  1     kg
```
etc

Then you need a way to define conversions between different standards.

So you want to have named standards (imperial, metric, etc) and equivalencies between them

```
grams  metric  ounces imperial
liters metric  cups imperial
 ```
And on any recipe page a quick dropdown to switch between different defined standards

Finally you want a way to define conversions between units with a direction, a minimum/maximum for conversion and a remainder.
```
grams  kg   >  2000  0
tsp    tbsp >  3     1
tbsp   cups >  6   .25
cups   tbsp <  .25   0
```
To explain that last table a little more, you start the conversion at the minimum. So if you have 1900 grams, no conversion. If you have 2000 grams, you'd convert to 2 kg. So for tsp, if you have 3 or more tsps, you convert to tbsp.

However the remainder also comes into play. So a remainder limit of zero means no limit, so in the case of grams, if you have 2050 grams, you get 2.05 kg. `tsp -> tbsp` comes with a limit of 1, which means no decimals. So if you have 3.5 tsp you get 1 tbsp + .5 tsp. If you have 4 tsp you get 1 tbsp + 1 tsp.

And then for tbsp to cups, you have a remainder of .25, which means anything that you convert to the nearest quarter cup and then leave the rest in tbsp. So 10 tbsp converts to .5 cup + 2 tbsp. 12 tbsp converts cleanly to .75 cups. Conversely, if you ever get cups below .25 cups, it will convert it to tbsp with no remainder limit, so .175 cups becomes 2.8 tbsp.

This last set of conversions (if defined/activated in mealie) I think would always be applied to recipes (and that would nicely solve the issues raised in 3127 I think), however you could add an option that those conversions are only applied if the units in a recipe are changed to a different standard or if the recipe is scaled. That means if you view the base recipe the units are always as they were initial designated. Personally, I think I'd always want those conversions applied. But tastes will likely vary.

**Reply by ArLo69** [reply]
I do indeed mean no hardcoded conversion. But a table that can be filled and altered by the user. A prefilled table would be nice indeed. That is why I talk about a table. My goal is to make one system that helps the user with all kinds of conversions in quantities and units.

So beside conversions between large and smaller units and different measuring systems people can also make their own conversion like 1 teaspoon = 2 grams (something a friend likes to do) or 1 cup = 20 teaspoons or whatever they like.

But my suggestion for the unit conversion makes more flexibility possible. I did 'steal' the idea from an ERP application that uses this system for over 30 years for a reason ;)
The records with an empty conversion are meant as base unit. And that helps to solve your problem.

The reason is this:
If you have two recipes and in one you use 1 pond (500 grams) and in another 4 ons (100 grams) then you have two different units.
Now mealie wants to make a shopping list and the vegatable is selled in kilograms. If wished by popular demand in the future, ingredients could hold the unit in which it is selled, what can be used to create shopping lists with actual quantities.
Then you need to calculate kilograms in my example.
In this example you would need conversions between ons and kilogram and pond and kilogram and with a more complex scale (ml, cl, dl, l)  many more.
Here you calculate 
1 pond = 500 gram
4 ons = 400 gram
400 + 500 = 900 gram
900 gram = 0.9 kg
And on the shopping list you will need to round that up to 1 kg as that is the smallest unit sold. In a recipe you could print it as 0.9 kg or as 9 ons (largest unit to give rounded numbers)

That is the use of this 'base unit' it makes calculations more easy. Especially when you have to rely on it inside one system. I know you allways have to use two steps, but for your example the flexible setup also helps:
You allready have .25 cup as your smallest unit. equal to 6 tbsp
If you link that back you 18 tsp (= 6 tbsp = 0.25 cup)
Say I have 13 tbsp. That would be 39 tsp or 0.5 cup and 1 tbsp
So:
13 tbsp = 39 tsp
Now comes the calculation in steps.
Find the largest unit with whole numbers:
0.25 cup = 18 tsp
1 cup = 72 tsp
39 is in betweem so you pick 0.25 cup.
2x18 = 36, so you find 2 x 0.25 cup
39 - 36 = 3
3 tsp = 1 tbsp
So in the end you find:
2 x 0.25 cup and 1 tbsp

If those 39 tsp were 40 then that would result in 
2 x 0.25 cup and 1 tbsp and 1 tsp

I hope this makes clear that though calculating back to the smallest unit in the conversion and then back up, it helps to solve the puzzle with roundings.

I would convert the units when I put the recipe in mealie. But as you suggest storing the originals and then a "show in system x" would be great. 
And also here the base unit might get the name of the system (metric, imperial, 'my own') and help mealie figure out how to calculate.

**Reply by crimsoncor** [reply]
I think we're mostly at semantic differences on the idea of empty units. Because when you've established chains of conversions
```
tsp < tbsp < cups
ml < cl < dl < c
```
the lowest unit in any chain of conversions is the empty unit in your example. I don't think you gain anything other than added UX complexity from requiring explicit empty units vs implicit ones.

**Reply by Ronaldvr** [reply]
In the Netherlands we have a baker who wrote a now quite famous cook book (Holtkamp https://watschaftdepodcast.com/kookboek/recensie-de-banketbakker-holtkamp-freud/ )  who *never* uses  volume measurements but actually gives weights for nearly everything in his recipes

**Comment by meowmachine** [comment]
If each ingredient has something like a "grams_per_cup", field in the database, those things could be calculated. 

Fortunately, the USDA food database has these values:
(foundation foods > json file)
https://fdc.nal.usda.gov/download-datasets

In the "foodPortions" nodes of that USDA database, it has "gramWeight "of the portion size, so you can derive something like "129 grams per 1 cup". 
So, as long as that weight-to-volume ratio is in the database for each ingredient, the imperial>metric conversions would be possible. It would then just be a matter of converting units.

A bonus of adding this value in the database would be that you can also use it to calculate nutrition facts for scaled portions because that same USDA database has a "calories per 100 grams" value too.

**Reply by fbd-ss** [reply]
the USDA will have to get with the program then

**Reply by meowmachine** [reply]
> the USDA will have to get with the program then

They already are with the program, what do you mean? They have the data. Mealie just has to implement it.

**Comment by EinToni** [comment]
A lot of people a mixing up units of volume with units of weight.
One cannot simply convert between cups and grams or tablespoons and grams, because gram is a measure of weight and the other two measure volume. The volume is strongly dependent on the ingretient one measures.
E.g. 1 tbsp water ~ 15g, but 1 tbsp flour ~ 8g (depends on the density, I could also stack way more on a tbsp).

We could assume generic defaults for those conversions, but that could become very inaccurate.
Better would be to define conversion per ingredient or use a source like USDA as already mentioned.

**Reply by tecosaur** [reply]
Yup. I think it's worth differentiating between _unit conversion_ (different measures for the same physical quality, e.g. mass) and _dimension conversion_ (e.g. going from volume to mass).

**Reply by meowmachine** [reply]
You don't have to assume generic conversions. The USDA food database already has all these conversions. For example they say exactly how milligrams are in 1 cup of flower. You could call it "food density".

(foundation foods > json file)
https://fdc.nal.usda.gov/download-datasets

Edit: Sorry I didn't see read your last sentence where you mentioned the USDA database.

**Reply by ArLo69** [reply]
I agree with you. And I had exacly the same thought with converting volumes and weights. But I know people who hate to work with volumes. I am not familiar with baking. But as I get it there weight is more important for a good mix where with cooking it depends more on taste. Since even not every tablespoon has the same size.

Anyway my idea was to get a configurable system that can be used at more places in mealie to convert units the way the users wants.

**Reply by fbd-ss** [reply]
every volume of an item has a weight.

**Comment by MaxKiv** [comment]
I'd love this feature

**Comment by eumel2** [comment]
Id love to see that implemented and then in addition use these metrics to automatically calculate nutrition as well. 
Im not even going so far as to say this needs to be shipped with all units and conversions and nutrition values already in, even just having the empty fields to fill out myself would be great.

**Comment by MRRandomniss** [comment]
This would really be great to have.

**Comment by ronnybugle** [comment]
I've been waiting on this feature for years.
Every single bloody recipe I import needs to be converted from US to metric manually and it's so tedious and I can't believe this isn't a feature already. Why hasn't this been worked on when this was opened years ago? If I want to work on it to help, how would I do that? It feels like otherwise it won't happen. I get the impression that maybe it hasn't been implemented yet because of a few sticking points, but surely anything that even does part of the job is a better start point than no helpful functionality?

**Comment by Xyolyp** [comment]
@ronnybugle asked upstream how to help move this along — there is already work in flight: #7573 by @agorton implements it (display-only, metric/imperial/US). It is a draft and has gone stale without a maintainer review, not for lack of code.

I have built a variant on top of it and left the details in a comment there, so the discussion stays in one place rather than two. Short version of where it stands and what is deliberately out of scope, since both come up repeatedly in this thread:

- **Volume never becomes weight.** Cups to grams needs a density per food, which Mealie does not store. @EinToni and @meowmachine are right that this is a separate feature with its own data model (the USDA portion data is one plausible source). Same-dimension conversion works today without any of that.
- **Temperature is a separate setting.** @Spinkman asked for grams with a Fahrenheit oven; that combination works.
- **The UK cup problem is real.** As @enrauk noted, UK/AU/US measures differ. Mealie standardizes every "cup" to the US cup (236.59 ml), so a UK mode cannot be fully correct until that is addressed — imperial therefore renders volumes below a pint in millilitres rather than inventing an imperial cup.

Anyone who wants to try it can point at the branch linked in #7573. Reviews and opinions on the unit-choice rules are more useful than more upvotes at this point.

**Comment by deanmalan** [comment]
I feel like, in this day and age, adding the option of an LLM-based unit conversion step, similar to the `TranslateRecipeStep` could be an option here. Ideally with the option of adding a custom prompt.

**Reply by deanmalan** [reply]
Thought about it now, and while I still think this would be the quickest win, it would be quite shite if your baking flops because an LLM converted the temperature incorrectly. 😢

**Comment by Xyolyp** [comment]
@deanmalan interesting thought, and it is the right question to ask now rather than later — but I would keep the two apart.

For same-dimension conversion an LLM is a heavy tool for a light job. Grams to ounces is eight units and two dimensions of arithmetic: it is exact, it reproduces, it runs offline, it costs nothing per recipe view, it needs no provider configured, and it cannot hallucinate 4.2 when the answer is 3.5. A round trip per recipe — or per toggle — would be a real regression in how the page feels, and it would put a feature that works for everyone today behind a service many self-hosters deliberately do not run.

Where an LLM would genuinely earn its keep is exactly where #8202 stops: **volume to weight**. "1 cup of flour" in grams depends on the ingredient, and that is knowledge, not arithmetic — the same gap @EinToni and @meowmachine describe above. That is worth building, and it is a different feature with a different failure mode: wrong-but-plausible instead of impossible.

So my suggestion would be to layer rather than replace: deterministic conversion as the baseline that always works, and an opt-in LLM step on top for the cases arithmetic cannot reach. If you want to build that, I would be glad to see it — and #8202 does not stand in its way, since ingredients it cannot convert are passed through untouched rather than guessed at.

**Reply by deanmalan** [reply]
I'd suggest reading https://dontbeameatproxy.com/ if you want others to engage with you in good faith.

**Reply by Xyolyp** [reply]
@deanmalan I do leverage LLMs a lot but it's assisting me and only rarely running autonomously. This specific comment for example is my thought just phrased with an LLM

**Reply by Xyolyp** [reply]
@deanmalan if this work style is not welcome I'm happy to adjust.

**Comment by michael-genson** [comment]
For anyone listening: currently we have https://github.com/mealie-recipes/mealie/pull/8292 open to support this. Would users prefer to be able to disable this feature for a household, or should it be generally available?

It has a low footprint, just one UI element:
<img width="585" height="348" alt="image" src="https://github.com/user-attachments/assets/5df69f13-227d-47f1-bc66-8727ca17fd17" />

**Reply by hilburn** [reply]
This is great to see, I just started using Mealie and the lack of this is already a little frustrating
I think the more useful option would be (preferably per user but household if that's problematic) - set a default on it. And I guess you could add a 4th "hide" option as well

**Reply by michael-genson** [reply]
The PR already remembers your last choice, so after you choose once it'll persist between sessions.

Currently there is no "hide" though.

**Reply by hilburn** [reply]
Ah, remembering is good for a per user basis definitely - I thought the hide option would be good if it was a selection box in the settings but if that's not there, then yeah maybe just a button to toggle it off?

Now the next thing will be customising the rounding - I'd want 14oz to convert to 400g, not 396! Maybe something like "nearest round number within X%"

**Reply by AlbinLind** [reply]
Personally, I see no reason to have to hide/disable it. It's a quite important feature, as a metric user who has to depend on a lot of US recipes. If you don't need to change it, you don't have to think about it. Similar as serving size

**Reply by bferd** [reply]
I'm Canadian so I flip-flop between Metric and US Imperial all the time. Hide (recipe) disable (household) choices would be ideal for me.

**Reply by hilburn** [reply]
@AlbinLind 
I can see people wanting to hide it just to keep the UI tidy - having an option you have to scroll past every recipe that you never use could be annoying

**Reply by AlbinLind** [reply]
@hilburn 
I see that point. I don't mind it being able to be disabled, but I think it would be odd to have it as an opt-in, instead of opt-out (as suggested). This is a feature that already exists on many recipe websites, and I was honestly surprised to learn that they didn't have it yet.

So, I think it is completely fine to hide it, but I, personally, do not find it necessary. I'm more invested in having the feature.

**Reply by hilburn** [reply]
Completely agree :)
